modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/ExtensionMessageSource.java

```diff
@@ -204,20 +217,41 @@ private synchronized void createSource(boolean restarting) throws Exception {
     }
   }
 
-  private void startSource(boolean restarting, RestartContext restartContext)
-      throws MuleException {
+  private void startSource(boolean restarting, RestartContext restartContext) throws MuleException {
+    Runnable onSuccess;
+    Consumer<Throwable> onFailure;
+    if (retryPolicyTemplate.isAsync()) {
+      onSuccess = this::onReconnectionSuccessful;
+      onFailure = this::onReconnectionFailed;
+    } else {
+      onSuccess = () -> {
+      };
+      onFailure = (t) -> {
+      };
+    }
+    Supplier<CompletableFuture<Void>> futureSupplier = () -> {
+      CompletableFuture<Void> future = new CompletableFuture<>();
+      retryScheduler.execute(() -> doWork(restarting, restartContext, future));
+      return future;
+    };
+    CompletableFuture<Void> future = retryPolicyTemplate
+        .applyPolicy(futureSupplier, t -> true, t -> computeStats(), NULL_THROWABLE_CONSUMER, identity(), retryScheduler)
+        .whenComplete((v, e) -> {
+          if (e != null) {
+            onFailure.accept(e);
+          } else {
+            onSuccess.run();
+          }
+        });
     try {
-      if (restarting && retryPolicyTemplate.isAsync()) {
```

Comment:
this looks like a fundamental departure of a behavior that was explicitly put in place to avoid race conditions. @fsgonz can you please review this? 

@SebaElizalde why is the restarting condition no longer relevant? Why is the IMMEDIATE_SCHEDULER no longer used?

Reply:
This has been change since introduced originally (MULE-17267). The idea then was to perform the retry in another thread in case of restarting (actually this was already changed in MULE-18777). Does this not fail for the app in the ticket associated to MULE-17267?

Reply:
There is no app associated to MULE-17267 sadly.
Indeed it was changed in MULE-18777: https://github.com/mulesoft/mule/commit/e19925827c5249e9435122c5fe2ab7b4f4bca45b (line 209), which makes it confusing as to what behaviour should be honoured. Perhaps line 314 of that PR causes that the reconnection to always executes in another thread regardless of the policy.
In any case, we could now handle sync/async with the future instead of the scheduler?

Reply:
We've verified that the thread switch in the case of reconnection happens before that, so it won't be a problem.